src/mailroom.py

```diff
@@ -0,0 +1,155 @@
+
+
+def amount_prompt_validator(donation_amount):
+    """Validate donation amount."""
+    if donation_amount == u'q':
+            main_function()
+    else:
+        try:
+            if int(donation_amount) or float(donation_amount):
+                return float(donation_amount)
+        except ValueError:
+            print(u'Invalid Input')
+            return u'invalid input'
+
+
+def update_dict(donor_name, donation_amount, dictionary):
+    """Check if name is in dict, add if not and add donation to the dict."""
+    dictionary.setdefault(donor_name, []).append(donation_amount)
+    return dictionary
+
+
+def generate_letter(donor_name, donation_amount):
+    """Generate thank you letter with name and new donation."""
+    print(u'Dear {0}, thank you for your donation of '
+          u'${1}.'.format(donor_name.capitalize(), donation_amount))
+    return (u'Dear {0}, thank you for your donation of '
+            u'${1}.'.format(donor_name.capitalize(), donation_amount))
+
+
+def do_math(dictionary):
+    """Find sum, no. of donations, and avg of donations for individuals."""
+    summary_list = []
+    for key in dictionary:
+        donor_math_list = []
+        value = dictionary[key]
+        total = sum(value)
+        count = len(value)
+        average = total / count
+        donor_math_list = [key, total, count, average]
+        summary_list.append(donor_math_list)
+    return sorted(summary_list, key=lambda x: x[1])
+
+
+def create_report(list_sorted):
+    """Print report of donors, with sum, count of, and avg of donations."""
+    print(u'{:<15} {:<15} {:<15} {:<15}'
+          .format(u'Name', u'Total', u'Count', u'Average'))
+    print_out_list = []
+    for individual_donor_list in list_sorted:
+        print_each_donor = (u'{:<15} {:<15} {:<15} {:<15}'
+                            .format(*individual_donor_list))
+        print_out_list.append(print_each_donor)
+        print(print_each_donor)
+    return print_out_list
+
+
+def dict_update_generate_letter(donor_name, dictionary):
+    """Update the dictionary and generate the Thank-You letter."""
+    donation_amount = amount_prompt_validator(amount_prompt())
+    while donation_amount == u'invalid input':
+        donation_amount = amount_prompt_validator(amount_prompt())
+    update_dict(donor_name, donation_amount, dictionary)
+    generate_letter(donor_name, donation_amount)
+
+
+def main_function():
+    """Main function for control flow"""
+    user_menu_choice = initial_prompt_validator(initial_prompt())
+    if user_menu_choice == u'l':
+        donor_name = name_prompt_validator(name_prompt())
+        while donor_name == u'l':
+            for key in DONOR_DICT:
+                print (key)
+            donor_name = name_prompt_validator(name_prompt())
+        while donor_name == u'invalid input':
+            donor_name = name_prompt_validator(name_prompt())
+        dict_update_generate_letter(donor_name, DONOR_DICT)
+        main_function()
```

Comment:
Do you need to recursively call this here?  Can you figure out how to return to the beginning of this function without making this recursive call?  You should.

The problem here is that you end up creating an ever-growing stack of call frames for each call to `main_function`.  Because none of them ever exit until you quit the entire program, you just keep using more and more memory.  That's not good.